src/v/cloud_storage/remote_segment.cc

```diff
@@ -177,6 +183,151 @@ remote_segment::data_stream(size_t pos, ss::io_priority_class io_priority) {
+                      uint64_t size_bytes,
+                      ss::input_stream<char> s) -> ss::future<uint64_t> {
+        offset_index tmpidx(get_base_rp_offset(), get_base_kafka_offset(), 0);
+        auto [sparse, sput] = input_stream_fanout<2>(std::move(s), 1);
+        auto parser = make_remote_segment_index_builder(
+          std::move(sparse),
+          tmpidx,
+          _base_offset_delta,
+          remote_segment_sampling_step_bytes);
+        auto fparse = parser->consume().finally(
+          [parser] { return parser->close(); });
+        auto fput = _cache.put(_path, sput).finally([sref = std::ref(sput)] {
+            return sref.get().close();
+        });
+        auto [rparse, rput] = co_await ss::when_all(
+          std::move(fparse), std::move(fput));
+        bool index_prepared = true;
+        if (rparse.failed()) {
+            vlog(
+              _ctxlog.warn,
+              "Failed to build a remote_segment index, error: {}",
+              rparse.get_exception());
+            rparse.ignore_ready_future();
+            index_prepared = false;
+        }
+        if (rput.failed()) {
+            vlog(
+              _ctxlog.warn,
+              "Failed to write a segment file to cache, error: {}",
+              rput.get_exception());
+            rput.get();
+        }
+        if (index_prepared) {
+            auto index_stream = make_iobuf_input_stream(tmpidx.to_iobuf());
+            co_await _cache.put(_path().native() + ".index", index_stream);
+            _index = std::move(tmpidx);
+        }
+        co_return size_bytes;
+    };
+
+    retry_chain_node local_rtc(
+      cache_hydration_timeout, cache_hydration_backoff, &_rtc);
+
+    auto res = co_await _api.download_segment(
+      _bucket, _path, callback, local_rtc);
+
+    if (res != download_result::success) {
+        vlog(
+          _ctxlog.debug,
+          "Failed to hydrating a segment {}, {} waiter will be "
+          "invoked",
+          _path,
+          _wait_list.size());
+        throw download_exception(res, _path);
+    }
+}
+
+ss::future<> remote_segment::maybe_materialize_index() {
+    ss::gate::holder guard(_gate);
+    auto path = _path().native() + ".index";
+    offset_index ix(_base_rp_offset, _base_rp_offset - _base_offset_delta, 0);
+    if (co_await _cache.is_cached(path) == cache_element_status::available) {
+        vlog(_ctxlog.info, "Trying to materialize index '{}'.", path);
+        auto cache_item = co_await _cache.get(path);
+        try {
+            ss::file_input_stream_options options{};
+            options.buffer_size
+              = config::shard_local_cfg().storage_read_buffer_size;
+            options.read_ahead
+              = config::shard_local_cfg().storage_read_readahead_count;
+            options.io_priority_class
+              = priority_manager::local().shadow_indexing_priority();
+            auto inp_stream = ss::make_file_input_stream(
+              cache_item->body, options);
```

Comment:
It is still possible that `cache_item` is empty here, and we need to bail out in this case.

Reply:
yep, going to fix this